optuna/_gp/_search_space.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from enum import IntEnum
+import math
+from typing import Any
+
+import numpy as np
+import scipy.stats.qmc
+
+import optuna
+
+
+class ParamType(IntEnum):
+    LINEAR = 0
+    LOG = 1
+    CATEGORICAL = 2
+
+
+@dataclass(frozen=True)
+class SearchSpace:
+    param_type: np.ndarray
+    bounds: np.ndarray
+    step: np.ndarray
+
+
+def untransform_one_param(
+    x: np.ndarray, param_type: ParamType, bounds: tuple[float, float], step: float
+) -> np.ndarray:
+    if param_type == ParamType.CATEGORICAL:
+        return x
+    else:
+        bounds2 = (bounds[0] - 0.5 * step, bounds[1] + 0.5 * step)
+        if param_type == ParamType.LOG:
+            bounds2 = (math.log(bounds2[0]), math.log(bounds2[1]))
+        x = x * (bounds2[1] - bounds2[0]) + bounds2[0]
+        if param_type == ParamType.LOG:
+            x = np.exp(x)
+        return x
+
+
+def transform_one_param(
+    x: np.ndarray, param_type: ParamType, bounds: tuple[float, float], step: float
+) -> np.ndarray:
+    if param_type == ParamType.CATEGORICAL:
+        return x
+    else:
+        bounds2 = (bounds[0] - 0.5 * step, bounds[1] + 0.5 * step)
+        if param_type == ParamType.LOG:
+            bounds2 = (math.log(bounds2[0]), math.log(bounds2[1]))
+            x = np.log(x)
+        x = (x - bounds2[0]) / (bounds2[1] - bounds2[0])
+        return x
+
+
+def round_one_transformed_param(
```

Comment:
I feel that "round" is more appropriate here, because in most cases this actually means to cut off numbers after the decimal point (but done in the transformed parameter space).

"Normalize" is too ambiguous, and it could mean a reversible process of transforming the parameter from the "unnormalized" space to a "normalized" one (which is incorrect here), or it could mean a process of translating  input (like a file path) into something equivalent but more canonical (which is again not a very accurate description of what this function does, because the input of this function is not equivalent to the output).

Reply:
Sorry, I do not remember why I left this comment in this line...
I was probably looking at the normalize function, but somehow commented here to rename it. Anyways, please ignore it!